src/pow.cpp

```diff
@@ -93,42 +92,22 @@ unsigned int CalculateNextWorkRequired(arith_uint256 bnAvg,
     return bnNew.GetCompact();
 }
 
-bool CheckEquihashSolution(const CBlockHeader *pblock, int nHeight, const Consensus::Params& params)
+bool CheckEquihashSolution(const CBlockHeader *pblock, const Consensus::Params& params)
 {
     unsigned int n = params.nEquihashN;
     unsigned int k = params.nEquihashK;
 
-    // Hash state
-    crypto_generichash_blake2b_state state;
-    EhInitialiseState(n, k, state);
-
     // I = the block header minus nonce and solution.
     CEquihashInput I{*pblock};
     // I||V
     CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
     ss << I;
 
-    // From Heartwood activation, check with the Rust validator
-    if (params.NetworkUpgradeActive(nHeight, Consensus::UPGRADE_HEARTWOOD)) {
-        return librustzcash_eh_isvalid(
-            n, k,
-            (unsigned char*)&ss[0], ss.size(),
-            pblock->nNonce.begin(), pblock->nNonce.size(),
-            pblock->nSolution.data(), pblock->nSolution.size());
-    }
-
-    // Before Heartwood activation, check with the C++ validator
-    ss << pblock->nNonce;
-
-    // H(I||V||...
-    crypto_generichash_blake2b_update(&state, (unsigned char*)&ss[0], ss.size());
-
-    bool isValid;
-    EhIsValidSolution(n, k, state, pblock->nSolution, isValid);
-    if (!isValid)
-        return error("CheckEquihashSolution(): invalid solution");
-
-    return true;
+    return librustzcash_eh_isvalid(
```

Comment:
For future changes of this sort, I think it would be a good idea to, for some period prior to the full cut-over between implementations, run *both* checks for some period to ensure that they obtained the same result. Although I guess that for this check it's likely either always correct or never correct, so perhaps unit testing is sufficient.

Reply:
That's what I originally implemented in #2851, but it was never merged, and given that `zebrad` will soon be using the same Rust validator as we will now be using, a switch coincident with a NU made sense. We could instead have treated it as an explicit (though no-op) consensus change, and we would have deployed it in the same manner.